src/audio/google/google_rtc_audio_processing.c

```diff
@@ -544,222 +546,311 @@ static int google_rtc_audio_processing_prepare(struct processing_module *mod,
-
-			if (cd->raw_mic_buffer_frame_index == cd->num_frames) {
-				GoogleRtcAudioProcessingProcessCapture_int16(cd->state,
-									     cd->raw_mic_buffer,
-									     cd->output_buffer);
-				cd->output_buffer_frame_index = 0;
-				cd->raw_mic_buffer_frame_index = 0;
-			}
+	source_release_data(src, bytes);
+}
 
-			src += audio_stream_get_channels(mic_stream);
-			dst += audio_stream_get_channels(out_stream);
+static void mic_out_copy(struct sof_sink *sink, int frames, float **src_bufs)
+{
+	size_t chan = MIN(MIC_CHAN_MAX, sink_get_channels(sink));
+	size_t samples = frames * chan;
+	size_t bytes = samples * sizeof(mic_sample_t);
+	mic_sample_t *buf, *bufstart, *bufend;
+	int i, c, err;
+	size_t bufsz;
+	float *src[MIC_CHAN_MAX];
+
+	for (i = 0; i < chan; i++)
+		src[i] = src_bufs[i];
+
+	err = sink_get_buffer(sink, bytes, (void *)&buf, (void *)&bufstart, &bufsz);
+	assert(err == 0);
+	bufend = &bufstart[bufsz];
+
+	for (i = 0; i < frames; i++) {
+		for  (c = 0; c < chan; c++) {
+			*buf++ = float_to_mic(*src[c]++);
+			if (buf >= bufend)
+				buf = bufstart;
 		}
-		num_frames_remaining -= n;
-		src = audio_stream_wrap(mic_stream, src);
-		dst = audio_stream_wrap(out_stream, dst);
 	}
+	sink_commit_buffer(sink, bytes);
+}
+
+static int mod_process(struct processing_module *mod, struct sof_source **sources,
+		       int num_of_sources, struct sof_sink **sinks, int num_of_sinks)
+{
+	struct google_rtc_audio_processing_comp_data *cd = module_get_private_data(mod);
+
+	if (cd->reconfigure)
+		google_rtc_audio_processing_reconfigure(mod);
+
+	struct sof_source *mic = sources[cd->raw_microphone_source];
+	struct sof_source *ref = sources[cd->aec_reference_source];
+	struct sof_sink *out = sinks[0];
+
+	bool ref_ok = cd->ref_comp_buffer->source->state == COMP_STATE_ACTIVE;
+
+	/* Would be cleaner to store a bit of state to elide a bzero
+	 * we already did, but we'd be doing the copy of real data in
+	 * the ref_ok state anyway.
+	 */
+	if (!ref_ok)
+		bzero(refoutbuf, sizeof(refoutbuf));
+
+	int fmic = source_get_data_frames_available(mic);
+	int fref = source_get_data_frames_available(ref);
+	int frames = ref_ok ? MIN(fmic, fref) : fmic;
+	int n, frames_rem;
+
+	/* If fref > fmic (common at pipeline startup if
+	 * playback was already active), we should consume the early
+	 * samples so AEC compares the most recent values.
+	 */
+	if (ref_ok && fref > fmic)
+		source_release_data(ref, (fref - fmic) * cd->ref_frame_bytes);
+
+	for (frames_rem = frames; frames_rem; frames_rem -= n) {
+		n = MIN(frames_rem, cd->num_frames - cd->buffered_frames);
 
-	module_update_buffer_position(&input_buffers[cd->raw_microphone_source],
-				      &output_buffers[0], frames);
+		int si = cd->buffered_frames * source_get_channels(mic);
+
+		mic_in_copy(mic, n, cd->raw_mic_buffers, si);
+
+		if (ref_ok)
+			ref_copy(ref, n, cd->refout_buffers, si);
+
+		cd->buffered_frames += n;
+
+		if (cd->buffered_frames >= cd->num_frames) {
+			execute_aec(cd);
```

Comment:
How do we garantee the ref have 10ms at the exact same time the capture have 10ms. Is there no risk of wrapped around buffer when they are out of sync ?